include/votca/xtp/bseoperator_btda.h

```diff
@@ -0,0 +1,235 @@
+/*
+ * Copyright 2009-2019 The VOTCA Development Team (http://www.votca.org)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+
+#ifndef __VOTCA_BSEOP_BTDA_H
+#define __VOTCA_BSEOP_BTDA_H
+#include <votca/xtp/bse_operator.h>
+#include <votca/xtp/eigen.h>
+namespace votca {
+namespace xtp {
+
+template<typename MatrixReplacementA, typename MatrixReplacementB>
+class HamiltonianOperator;
+
+}
+}  // namespace votca
+namespace Eigen {
+namespace internal {
+// MatrixReplacement looks-like a SparseMatrix, so let's inherits its traits:
+template <typename MatrixReplacementA, typename MatrixReplacementB>
+struct traits<votca::xtp::HamiltonianOperator<MatrixReplacementA,MatrixReplacementB>>
+    : public Eigen::internal::traits<Eigen::MatrixXd> {};
+}  // namespace internal
+}  // namespace Eigen
+
+namespace votca {
+namespace xtp {
+template <typename MatrixReplacementA, typename MatrixReplacementB>
+class HamiltonianOperator
+    : public Eigen::EigenBase<HamiltonianOperator<MatrixReplacementA,MatrixReplacementB>> {
+ public:
+  // Required typedefs, constants, and method:
+  typedef double Scalar;
+  typedef double RealScalar;
+  typedef int StorageIndex;
+  enum {
+    ColsAtCompileTime = Eigen::Dynamic,
+    MaxColsAtCompileTime = Eigen::Dynamic,
+    IsRowMajor = false
+  };
+
+  HamiltonianOperator(const MatrixReplacementA &A,
+                      const MatrixReplacementB &B) 
+  : _A(A), _B(B) 
+  { 
+    _size = 2*A.cols() ;
+    _diag = get_diagonal();
+   };
+      
+    
+  class InnerIterator {
+   public:
+    InnerIterator(const HamiltonianOperator& xpr, const Eigen::Index& id)
+        : _xpr(xpr), _id(id){};
+
+    InnerIterator& operator++() {
+      _row++;
+      return *this;
+    }
+    operator bool() const {
+      return _row < _xpr._size;
```

Comment:
? I do not understand the comment

Reply:
well that's yours, I've copied that from the dipole-dipole interaction class in `dipoledipoleinteraction.h`

Reply:
ahh hmm ups, That was